zerver/actions/message_send.py

```diff
@@ -1523,6 +1524,57 @@ def validate_stream_id_with_pm_notification(
     return stream
 
 
+def check_can_send_private_message(
+    realm: Realm, sender: UserProfile, user_profiles: Sequence[UserProfile], recipient: Recipient
+) -> None:
+    if sender.is_bot:
+        return
+
+    for user_profile in user_profiles:
+        if not (user_profile.is_bot or user_profile.id == sender.id):
+            break
+    else:
+        return
+
+    direct_message_permission_group = realm.direct_message_permission_group
+    if (
+        not hasattr(direct_message_permission_group, "named_user_group")
+        or direct_message_permission_group.named_user_group.name != SystemGroups.EVERYONE
+    ):
+        allowed_user_ids = get_user_group_member_ids(direct_message_permission_group)
+        if sender.id not in allowed_user_ids:
+            for user_profile in user_profiles:
+                if user_profile.id in allowed_user_ids:
+                    break
+            else:
+                raise JsonableError(
+                    _("You are not allowed to send direct messages to these recipient(s).")
+                )
```

Comment:
I think this might do multiple database queries unnecessarily. I guess `is_user_in_group()` isn't right, because that checks a single user; you might need a prep commit to add a variant `is_any_user_in_group(user_group, user_ids)` with very similar code to the existing one; can be added with a nice set of tests.

Reply:
I tested this - it only takes one query in the `get_user_group_member_ids` function.

Reply:
I tested this - checking `direct_message_permission_group` takes only 1 extra query.

Reply:
It's one database query, but it's an expensive one! We should not be calling `get_user_group_member_ids` to fetch the entire set of users who can send DMs, which might be 30k users in a large organization, since it's not relevant; we just need to know whether any of the 4 users involved in this particular DM thread have the permission.